### PR TITLE
Add IsEveryoneHere Bot Function

### PR DESCRIPTION
**Pull request by MabroukTBN**
*IsEveryoneHere* is a slash function that checks for:
- *Full Time* members
- That are working today (i.e. no approved leave)
- Are currently ***not*** in the Meeting Channel

Names of members that fit these checks will be sent to the author in a private message, otherwise the author will be notified that 'Eveyone is here'

Added *MeetingChannel_id* to .env.dist

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -79,4 +79,24 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
...
     await ctx.send(content = "Done", delete_after = 0.1)
 
+@slash.slash(name="IsEveryoneHere", description="Checks if all working 'Full Time' members are in voice channel", guild_ids=guild_ids)
+async def IsEveryoneHere(ctx):
+
+    guild = client.guilds[0]
+    voice_channel = client.get_channel(int(os.getenv("MeetingChannel_id")))
+    fulltime_role = discord.utils.get(guild.roles, name = "Full Time")
+    fulltime_member_list = list(filter(lambda user: fulltime_role in user.roles, guild.members))
+    voicechannel_fulltime_member_list = list(filter(lambda user : fulltime_role in user.roles, voice_channel.members))
```
Comment (AntWong-TheBadNomads) [inline main.py:89]:
```suggestion
    fulltime_member_list_in_voicechannel = list(filter(lambda user : fulltime_role in user.roles, voice_channel.members))
```

File: main.py [hunk 2]
```diff
@@ -79,4 +79,24 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
...
+
+    not_here = list(set(fulltime_member_list) - set(voicechannel_fulltime_member_list))
+    not_here = list(filter(lambda user : (leave_interface.IsMemberWorking(user.id, datetime.today()))[0], not_here))
```
Comment (AntWong-TheBadNomads) [inline main.py:92]: Can you create a card in to do to deal with departments that don't need to attend core please, when that's done, link that here.

Comment (MabroukTBN) [reply main.py:92]: Card added

File: main.py [hunk 3]
```diff
@@ -79,6 +79,34 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
...
+    approved_leaves_names = [member.display_name for member in approved_leaves]
+    approved_leaves_reasons = [leave_interface.IsMemberOnLeave(member.id, datetime.today())[1] for member in approved_leaves]
+    approved_leaves_dict = dict(zip(approved_leaves_names, approved_leaves_reasons))
+
+    missing_members = list(filter(lambda member : not (leave_interface.IsMemberOnLeave(member.id, datetime.today())[0]), not_here))
+    missing_members = [member.display_name for member in missing_members]
+
+    embed = UI.CreateIsEveryoneHereEmbed(approved_leaves_dict, missing_members, Utilities.IsAdmin(ctx.author))
+
+    if embed != None:
```
Comment (AntWong-TheBadNomads) [inline main.py:103]: this is now not needed. please don't rush your work

Comment (MabroukTBN) [reply main.py]: removed the useless code

File: UI.py [hunk 4]
```diff
@@ -208,6 +208,27 @@ def CreateIsMemberWorkingOptions():
...
+        embed.add_field(name = "Names", value = '\n'.join(missing_members), inline = False)
+    else:
+        embed.add_field(name = "Everyone's Here", value = "-", inline = False)
+
+    if approved_dict:
+        embed.add_field(name = "Approved Leaves", value = '\n'.join(approved_dict.keys()), inline = True)
+        if isAdmin:
+            embed.add_field(name = "Reasons", value = '\n'.join(approved_dict.values()), inline = True)
+
+    embed.set_footer(text = datetime.date.today())
```
Comment (AbdoHamed-TheBadNomads) [inline UI.py:230]:
```suggestion
    footer_text = (("\u200B " * embed_footer_spaces_count) + datetime.date.today().strftime("%d/%m/%Y"))
    embed.set_footer(text = footer_text)
```

File: main.py [hunk 5]
```diff
@@ -79,6 +79,33 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
 
     await ctx.send(content = "Done", delete_after = 0.1)
 
+@slash.slash(name="IsEveryoneHere", description="Checks if all working 'Full Time' members are in voice channel", guild_ids=guild_ids)
```
Comment (AbdoHamed-TheBadNomads) [inline main.py:82]:
```suggestion
@slash.slash(name="IsEveryoneHere", description="Checks if all working 'Full Time' members are in the meeting channel", guild_ids=guild_ids)
```

Comment (AbdoHamed-TheBadNomads) [reply main.py:82]: also spaces around `=`

File: main.py [hunk 6]
```diff
@@ -79,6 +79,33 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
...
+    today = datetime.today().date()
+    today = datetime(today.year, today.month, today.day)
+
+    guild = client.guilds[0]
+    meeting_channel = client.get_channel(int(os.getenv("MeetingChannel_id")))
+    fulltime_role = discord.utils.get(guild.roles, name = "Full Time")
+    
+    fulltime_members = list(filter(lambda member: fulltime_role in member.roles, guild.members))
+    fulltime_members_in_voicechannel = list(filter(lambda member : fulltime_role in member.roles, meeting_channel.members))
+    not_here = list(set(fulltime_members) - set(fulltime_members_in_voicechannel))
```
Comment (AbdoHamed-TheBadNomads) [inline main.py:94]: this can be done in a separate function `GetMissingMembers(role, channel)`

File: main.py [hunk 7]
```diff
@@ -79,6 +79,33 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
...
+
+    approved_leaves = list(filter(lambda member : (leave_interface.IsMemberOnLeave(member.id, today))[0], not_here))
+    approved_leaves_names = [member.display_name for member in approved_leaves]
+    approved_leaves_reasons = [leave_interface.IsMemberOnLeave(member.id, today)[1] for member in approved_leaves]
+    approved_leaves_dict = dict(zip(approved_leaves_names, approved_leaves_reasons))
```
Comment (AbdoHamed-TheBadNomads) [inline main.py:99]: this can be done in a separate function `GetApprovedLeavesForDay(member_list, day)`

File: main.py (hunk 6 above)
Comment (MabroukTBN) [reply main.py:94]: I initially had it as a separate function but couldn't find an appropriate place for it without cyclic imports. After speaking with Anthony about it we decided to keep it in main as is.

File: main.py (hunk 7 above)
Comment (MabroukTBN) [reply main.py:99]: I initially had it as a separate function but couldn't find an appropriate place for it without cyclic imports. After speaking with Anthony about it we decided to keep it in main as is.

File: main.py (hunk 5 above)
Comment (MabroukTBN) [reply main.py:82]: Done

File: main.py [hunk 8]
```diff
@@ -79,9 +79,12 @@ async def IsMemberWorking(ctx, discorduser, date = datetime.today().strftime('%d
 
     await ctx.send(content = "Done", delete_after = 0.1)
 
-@slash.slash(name="IsEveryoneHere", description="Checks if all working 'Full Time' members are in voice channel", guild_ids=guild_ids)
+@slash.slash(name="IsEveryoneHere", description = "Checks if all working 'Full Time' members are in the meeting channel", guild_ids = guild_ids)
 async def IsEveryoneHere(ctx):
 
+    today = datetime.today().date()
+    today = datetime(today.year, today.month, today.day)
```
Comment (AntWong-TheBadNomads) [inline main.py:86]: are you doing this to set the time to 12 am? there should be a better way to do so without reassigning a variable (which you seem to do often, try to reduce that)

Comment (MabroukTBN) [reply main.py:86]: This is done to zero out the time part of the datetime object. When using datetime.today(), you get values for the time that cause an error when comparing to the dates in the database.

Comment (AntWong-TheBadNomads) [reply main.py:86]: Read my comment again please. 
> there should be a better way to do so

Comment (MabroukTBN) [reply main.py:86]: This is an item I wanted to talk about for the upcoming sprint. Making the use of datetime/date consistent across all files.
For now would it be preferable to make a utility function to zero out the time from datetime objects?

Comment (AntWong-TheBadNomads) [reply main.py:86]: just checked, seems like `date` objects are slated to get a `.datetime()` function soon. this is fine for now

**Pull request merged**